data_ingestion/common/publisher.py

```diff
@@ -0,0 +1,25 @@
+from google.cloud import pubsub_v1
+
+
+def notify_data_ingested(id, gcs_bucket, filename):
```

Comment:
Whenever something is added/updated/deleted in a GCS bucket, you can configure a pubsub message to be emitted by GCS. We should be able to subscribe to those messages. https://cloud.google.com/storage/docs/pubsub-notifications

Reply:
The first thing I got working was using the finalize event from the GCS bucket: https://pantheon.corp.google.com/functions/details/us-central1/aaron-test-gcs-to-bq?project=fellowship-test-internal

I switched to this approach because Jennie mentioned a use case where the ingestion writes a bunch of different files (one per year) but wants to write them all to the same BQ table, so she didn't want to have one function triggered for each file.

Is this notification rule a third way of doing it?

Reply:
This would be using the finalize event. 

I'm wondering whether we need to ingest the yearly data every time. Or can we just do it once and then only ingest new years? Doing this only once and then ingesting the current or newest year available is less duplicate data transferred and then we can rely on the finalize even which is cleaner.

Reply:
I'm inclined to go with this approach for now, for velocity purposes since this is just a prototype. The way this is set up makes it pretty flexible while prototyping. Should be easy enough to change later.

The finalize event is probably cleaner if it covers our use cases, but I don't think it's so much cleaner that we should design our ingestion around it, especially since we don't know what all of our data sources will look like.

Do you think it makes sense to open github issues for things like this that we want to re-visit later on?

Reply:
Sounds good to me.

Re github issues: That seems like a good idea! I think we will need to revisit our ingestion strategy in general. Right now it makes sense to just copy everything all the time and dedupe later for velocity's sake, as you mentioned. But it can start getting expensive (there are credits for now but unsure how long they will last) and its nice to use fewer resources when possible without too much work.

Reply:
Makes sense. I'll file an issue.